File: modules/solid_mechanics/neml2/src/LAROMANCE6DInterpolation.cxx
//* This file is part of the MOOSE framework
//* https://www.mooseframework.org
//*
//* All rights reserved, see COPYRIGHT for full restrictions
//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
//*
//* Licensed under LGPL 2.1, please see LICENSE for details
//* https://www.gnu.org/licenses/lgpl-2.1.html

#include "LAROMANCE6DInterpolation.h"
#include <fstream>
#include <initializer_list>
#include "neml2/misc/math.h"

namespace neml2
{
register_NEML2_object(LAROMANCE6DInterpolation);

OptionSet
LAROMANCE6DInterpolation::expected_options()
{
  auto options = Model::expected_options();
  options.doc() = "Multilinear interpolation over six dimensions (von_mises_stress, temperature, "
                  "equivalent_plastic_strain, cell_dd_density, wall_dd_density, env_factor)";
  // Model inputs
  options.set<VariableName>("equivalent_plastic_strain") = VariableName("state", "ep");
  options.set<VariableName>("von_mises_stress") = VariableName("state", "s");

  options.set<VariableName>("cell_dd_density") = VariableName("old_state", "cell_dd");
  options.set<VariableName>("wall_dd_density") = VariableName("old_state", "wall_dd");

  options.set<VariableName>("temperature") = VariableName("forces", "T");
  options.set<VariableName>("env_factor") = VariableName("forces", "env_fac");
  // Model Outputs
  options.set_output("output_rate");
  // JSON
  options.set<std::string>("model_file_name");
  options.set<std::string>("model_file_variable_name");
  // jit does not currently work with this
  options.set<bool>("jit") = false; // false;
  options.set("jit").suppressed() = true;
  return options;
}

LAROMANCE6DInterpolation::LAROMANCE6DInterpolation(const OptionSet & options)
  : Model(options),
    _s(declare_input_variable<Scalar>("von_mises_stress")),
    _T(declare_input_variable<Scalar>("temperature")),
    _ep(declare_input_variable<Scalar>("equivalent_plastic_strain")),
    _cell_dd(declare_input_variable<Scalar>("cell_dd_density")),
    _wall_dd(declare_input_variable<Scalar>("wall_dd_density")),
    _env_fac(declare_input_variable<Scalar>("env_factor")),
    _output_rate(declare_output_variable<Scalar>("output_rate"))
{
  std::string filename = options.get<std::string>("model_file_name");
  std::ifstream model_file(filename.c_str());
  model_file >> _json;

  // storing grid points for indexing.
  // these should be stored differently so that they are all read in at once.  The order of this can
  // get messed up easily
  _stress_grid = json_vector_to_torch("in_stress");
  _temperature_grid = json_vector_to_torch("in_temperature");
  _plastic_strain_grid = json_vector_to_torch("in_plastic_strain");
  _cell_grid = json_vector_to_torch("in_cell");
  _wall_grid = json_vector_to_torch("in_wall");
  _env_grid = json_vector_to_torch("in_env");

  // Read in grid axes transform enums
  _stress_transform_enum = get_transform_enum(json_to_string("in_stress_transform_type"));
  _temperature_transform_enum = get_transform_enum(json_to_string("in_temperature_transform_type"));
  _plastic_strain_transform_enum =
      get_transform_enum(json_to_string("in_plastic_strain_transform_type"));
  _cell_transform_enum = get_transform_enum(json_to_string("in_cell_transform_type"));
  _wall_transform_enum = get_transform_enum(json_to_string("in_wall_transform_type"));
  _env_transform_enum = get_transform_enum(json_to_string("in_env_transform_type"));

  // Read in grid axes transform values
  _stress_transform_values = json_to_vector("in_stress_transform_values");
  _temperature_transform_values = json_to_vector("in_temperature_transform_values");
  _plastic_strain_transform_values = json_to_vector("in_plastic_strain_transform_values");
  _cell_transform_values = json_to_vector("in_cell_transform_values");
  _wall_transform_values = json_to_vector("in_wall_transform_values");
  _env_transform_values = json_to_vector("in_env_transform_values");

  // Storing values for interpolation
  _output_rate_name = options.get<std::string>("model_file_variable_name");
  _grid_values = json_6Dvector_to_torch(_output_rate_name);

  // set up output transforms
  if (_output_rate_name == "out_ep")
  {
    _output_transform_enum = get_transform_enum(json_to_string("out_strain_rate_transform_type"));
    _output_transform_values = json_to_vector("out_strain_rate_transform_values");
  }
  else if (_output_rate_name == "out_cell")
  {
    _output_transform_enum = get_transform_enum(json_to_string("out_cell_rate_transform_type"));
    _output_transform_values = json_to_vector("out_cell_rate_transform_values");
  }
  else if (_output_rate_name == "out_wall")
  {
    _output_transform_enum = get_transform_enum(json_to_string("out_wall_rate_transform_type"));
    _output_transform_values = json_to_vector("out_wall_rate_transform_values");
  }
  else
  {
    throw NEMLException("This ouput variable is not implemented, model_file_variable_name: " +
                        std::string(_output_rate_name));
  }
}

void
LAROMANCE6DInterpolation::request_AD()
{
  // only using first derivatives of out_ep, not out_cell and out_wall
  if (_output_rate_name == "out_ep")
  {
    std::vector<const VariableBase *> inputs = {&_s};
    _output_rate.request_AD(inputs);
  }
}

void
LAROMANCE6DInterpolation::set_value(bool out, bool dout_din, bool d2out_din2)
{
  neml_assert_dbg(!dout_din || !d2out_din2,
                  "Only AD derivatives are currently supported for this model");
  if (out)
    _output_rate = interpolate_and_transform();
}

LAROMANCE6DInterpolation::TransformEnum
LAROMANCE6DInterpolation::get_transform_enum(const std::string & name) const
{
  if (name == "COMPRESS")
    return TransformEnum::COMPRESS;
  else if (name == "DECOMPRESS")
    return TransformEnum::DECOMPRESS;
  else if (name == "LOG10BOUNDED")
    return TransformEnum::LOG10BOUNDED;
  else if (name == "EXP10BOUNDED")
    return TransformEnum::EXP10BOUNDED;
  else if (name == "MINMAX")
    return TransformEnum::MINMAX;

  throw NEMLException("Unrecognized transform: " + std::string(name));
}

std::pair<Scalar, Scalar>
LAROMANCE6DInterpolation::findLeftIndexAndFraction(const Scalar & grid,
                                                          const Scalar & interp_points)
{
  // idx is for the left grid point.
  // searchsorted returns the right idx so -1 makes it the left
  auto left_idx = torch::searchsorted(grid, interp_points) - 1;

  // this allows us to extrapolate
  left_idx = torch::clamp(left_idx, 0, grid.sizes()[0] - 2);

  Scalar left_coord = grid.index({left_idx});
  Scalar right_coord = grid.index({left_idx + torch::tensor(1, default_integer_tensor_options())});
  Scalar left_fraction = (right_coord - interp_points) / (right_coord - left_coord);

  return std::pair<Scalar, Scalar>(left_idx, torch::stack({left_fraction, 1 - left_fraction}, -1));
}

Scalar
LAROMANCE6DInterpolation::compute_interpolation(
    const std::vector<std::pair<Scalar, Scalar>> index_and_fraction, const Scalar grid_values)
{
  Scalar result = Scalar::zeros_like(_T);
  for (const auto i : {0, 1})
    for (const auto j : {0, 1})
      for (const auto k : {0, 1})
        for (const auto l : {0, 1})
          for (const auto m : {0, 1})
            for (const auto n : {0, 1})
            {

              auto vertex_value =
                  grid_values.index({(index_and_fraction[0].first +
                                      torch::tensor(i, default_integer_tensor_options())),
                                     (index_and_fraction[1].first +
                                      torch::tensor(j, default_integer_tensor_options())),
                                     (index_and_fraction[2].first +
                                      torch::tensor(k, default_integer_tensor_options())),
                                     (index_and_fraction[3].first +
                                      torch::tensor(l, default_integer_tensor_options())),
                                     (index_and_fraction[4].first +
                                      torch::tensor(m, default_integer_tensor_options())),
                                     (index_and_fraction[5].first +
                                      torch::tensor(n, default_integer_tensor_options()))});
              auto weight = index_and_fraction[0].second.select(-1, i) *
                            index_and_fraction[1].second.select(-1, j) *
                            index_and_fraction[2].second.select(-1, k) *
                            index_and_fraction[3].second.select(-1, l) *
                            index_and_fraction[4].second.select(-1, m) *
                            index_and_fraction[5].second.select(-1, n);
              result += vertex_value * weight;
            }
  return result;
}

/// compute interpolated value
Scalar
LAROMANCE6DInterpolation::interpolate_and_transform()
{
  // These transform constants should be given in the json file.
  auto cell_dd_transformed = transform_data(_cell_dd, _cell_transform_values, _cell_transform_enum);
  auto wall_dd_transformed = transform_data(_wall_dd, _wall_transform_values, _wall_transform_enum);
  auto s_transformed = transform_data(_s, _stress_transform_values, _stress_transform_enum);
  auto ep_transformed =
      transform_data(_ep, _plastic_strain_transform_values, _plastic_strain_transform_enum);
  auto T_transformed =
      transform_data(_T, _temperature_transform_values, _temperature_transform_enum);
  auto env_fac_transformed = transform_data(_env_fac, _env_transform_values, _env_transform_enum);

  std::vector<std::pair<Scalar, Scalar>> left_index_weight;
  left_index_weight.push_back(findLeftIndexAndFraction(_stress_grid, s_transformed));
  left_index_weight.push_back(findLeftIndexAndFraction(_temperature_grid, T_transformed));
  left_index_weight.push_back(findLeftIndexAndFraction(_plastic_strain_grid, ep_transformed));
  left_index_weight.push_back(findLeftIndexAndFraction(_cell_grid, cell_dd_transformed));
  left_index_weight.push_back(findLeftIndexAndFraction(_wall_grid, wall_dd_transformed));
  left_index_weight.push_back(findLeftIndexAndFraction(_env_grid, env_fac_transformed));
  Scalar interpolated_result = compute_interpolation(left_index_weight, _grid_values);
  Scalar transformed_result =
      transform_data(interpolated_result, _output_transform_values, _output_transform_enum);
  return transformed_result;
}

Scalar
LAROMANCE6DInterpolation::transform_data(const Scalar & data,
                                                const std::vector<Real> & param,
                                                TransformEnum transform_type) const
{
  switch (transform_type)
  {
    case TransformEnum::COMPRESS:
      return transform_compress(data, param);

    case TransformEnum::DECOMPRESS:
      return transform_decompress(data, param);

    case TransformEnum::LOG10BOUNDED:
      return transform_log10_bounded(data, param);

    case TransformEnum::EXP10BOUNDED:
      return transform_exp10_bounded(data, param);

    case TransformEnum::MINMAX:
      return transform_min_max(data, param);
  }
}

Scalar
LAROMANCE6DInterpolation::transform_compress(const Scalar & data,
                                                    const std::vector<Real> & param) const
{
  Real factor = param[0];
  Real compressor = param[1];
  Real original_min = param[2];
  auto d1 = math::sign(data) * math::pow(math::abs(data * factor), compressor);
  auto transformed_data = math::log10(1.0 + d1 - original_min);
  return transformed_data;
}

Scalar
LAROMANCE6DInterpolation::transform_decompress(const Scalar & data,
                                                      const std::vector<Real> & param) const
{
  Real factor = param[0];
  Real compressor = param[1];
  Real original_min = param[2];
  auto d1 = math::pow(10.0, data) - 1.0 + original_min;
  auto transformed_data = math::sign(d1) * math::pow(math::abs(d1), 1.0 / compressor) / factor;
  return transformed_data;
}

Scalar
LAROMANCE6DInterpolation::transform_log10_bounded(const Scalar & data,
                                                         const std::vector<Real> & param) const

{
  Real factor = param[0];
  Real lowerbound = param[1];
  Real upperbound = param[2];
  Real logmin = param[3];
  Real logmax = param[4];
  Real range = upperbound - lowerbound;
  auto transformed_data =
      range * (math::log10(data + factor) - logmin) / (logmax - logmin) + lowerbound;
  return transformed_data;
}

Scalar
LAROMANCE6DInterpolation::transform_exp10_bounded(const Scalar & data,
                                                         const std::vector<Real> & param) const
{
  Real factor = param[0];
  Real lowerbound = param[1];
  Real upperbound = param[2];
  Real logmin = param[3];
  Real logmax = param[4];
  Real range = upperbound - lowerbound;
  auto transformed_data =
      (math::pow(10.0, ((data - lowerbound) * (logmax - logmin) / range) + logmin) - factor);
  return transformed_data;
}

Scalar
LAROMANCE6DInterpolation::transform_min_max(const Scalar & data,
                                                   const std::vector<Real> & param) const
{
  Real data_min = param[0];
  Real data_max = param[1];
  Real scaled_min = param[2];
  Real scaled_max = param[3];
  auto transformed_data =
      ((data - data_min) / (data_max - data_min)) * (scaled_max - scaled_min) + scaled_min;
  return transformed_data;
}

std::string
LAROMANCE6DInterpolation::json_to_string(std::string key)
{
  if (!_json.contains(key))
    throw NEMLException("The key '" + std::string(key) + "' is missing from the JSON data file.");

  std::string name = _json[key].template get<std::string>();
  return name;
}

std::vector<Real>
LAROMANCE6DInterpolation::json_to_vector(std::string key)
{
  if (!_json.contains(key))
    throw NEMLException("The key '" + std::string(key) + "' is missing from the JSON data file.");

  std::vector<Real> data_vec = _json[key].template get<std::vector<Real>>();
  return data_vec;
}

Scalar
LAROMANCE6DInterpolation::json_vector_to_torch(std::string key)
{
  if (!_json.contains(key))
    throw NEMLException("The key '" + std::string(key) + "' is missing from the JSON data file.");

  std::vector<Real> in_data = _json[key].template get<std::vector<Real>>();
  const int64_t data_dim = in_data.size();
  return Scalar(torch::from_blob(in_data.data(), {data_dim}).clone());
}

Scalar
LAROMANCE6DInterpolation::json_6Dvector_to_torch(std::string key)
{
  using std::vector;
  if (!_json.contains(key))
    throw NEMLException("The key '" + std::string(key) + "' is missing from the JSON data file.");

  vector<vector<vector<vector<vector<vector<Real>>>>>> out_data =
      _json[key].template get<vector<vector<vector<vector<vector<vector<Real>>>>>>>();

  const int64_t sz_l0 = out_data.size();
  const int64_t sz_l1 = out_data[0].size();
  const int64_t sz_l2 = out_data[0][0].size();
  const int64_t sz_l3 = out_data[0][0][0].size();
  const int64_t sz_l4 = out_data[0][0][0][0].size();
  const int64_t sz_l5 = out_data[0][0][0][0][0].size();

  auto check_level_size =
      [](const int64_t current_vec_size, const int64_t sz_level, const std::string & key)
  {
    if (current_vec_size != sz_level)
      throw NEMLException("Incorrect JSON interpolation grid size for '" + key + "'.");
  };

  std::vector<Real> linearize_values;
  check_level_size(out_data.size(), sz_l0, key);
  for (auto && level1 : out_data)
  {
    check_level_size(level1.size(), sz_l1, key);
    for (auto && level2 : level1)
    {
      check_level_size(level2.size(), sz_l2, key);
      for (auto && level3 : level2)
      {
        check_level_size(level3.size(), sz_l3, key);
        for (auto && level4 : level3)
        {
          check_level_size(level4.size(), sz_l4, key);
          for (auto && level5 : level4)
          {
            check_level_size(level5.size(), sz_l5, key);
            for (auto && value : level5)
              linearize_values.push_back(value);
          }
        }
      }
    }
  }

  return Scalar(
      torch::from_blob(linearize_values.data(), {sz_l0, sz_l1, sz_l2, sz_l3, sz_l4, sz_l5})
          .clone());
}

} // namespace neml2

File: modules/solid_mechanics/data/laromance/LAROMANCE6DInterpolation.py
# This file is part of the MOOSE framework
# https://mooseframework.inl.gov
#
# All rights reserved, see COPYRIGHT for full restrictions
# https://github.com/idaholab/moose/blob/master/COPYRIGHT
#
# Licensed under LGPL 2.1, please see LICENSE for details
# https://www.gnu.org/licenses/lgpl-2.1.html

"""MOOSE-hosted NEML2 model: LAROMANCE 6D multilinear interpolation grid.

This is a NEML2 model authored and shipped by MOOSE (solid_mechanics). It is
imported into the embedded cpp-eager interpreter via a NEML2Action ``load``
parameter, which resolves this file on the application's data search path. Python
port of the v2 C++ ``LAROMANCE6DInterpolation`` -- a reduced-order creep model
that samples the inelastic strain rate and the cell/wall dislocation density
rates from a six-dimensional rectilinear grid (read from a JSON file) by
multilinear interpolation, with per-axis input transforms and an output
transform.

The grid axes, in order, are
``[von_mises_stress, temperature, equivalent_plastic_strain,
cell_dislocation_density, wall_dislocation_density, env_factor]``; the schema
declares the six model inputs plus one output rate selected by
``model_file_variable_name`` (``out_ep`` / ``out_cell`` / ``out_wall``).

For ``out_ep`` the model declares ``request_AD`` for ``d(output_rate) /
d(von_mises_stress)`` only -- mirroring the v2 model, which requested that single
first derivative for the implicit radial-return solve. The leaf therefore writes
only a value ``forward`` (no ``v=`` branch); the framework supplies the chain
rule by reverse-mode autodiff. request_AD is fully supported in cpp-eager (which
is what these tests use); the cell/wall instances are forward-only.

The grids are registered as buffers so they track the model's working device and
dtype under ``.to()`` (NEML2 builds typed tensors at float64) and so the eager
surface infers the model dtype from them.
"""

from __future__ import annotations

import itertools
import json
from typing import cast

import torch

from neml2.factory import register_neml2_object
from neml2.models.model import Model
from neml2.schema import HitSchema, input, option, output
from neml2.types import Scalar

# Grid axes, in interpolation order. The forward inputs are reordered onto this
# axis order before interpolation; the 6D grid values are indexed in this order.
_GRID_KEYS = ("in_stress", "in_temperature", "in_plastic_strain", "in_cell", "in_wall", "in_env")


@register_neml2_object("LAROMANCE6DInterpolation")
class LAROMANCE6DInterpolation(Model):
    """Multilinear interpolation over six dimensions.

    Inputs (von_mises_stress, temperature, equivalent_plastic_strain,
    cell_dislocation_density, wall_dislocation_density, env_factor) and a single
    output rate selected from the JSON grid by ``model_file_variable_name``.
    """

    hit = HitSchema(
        input("equivalent_plastic_strain", Scalar, "The equivalent plastic strain"),
        input("von_mises_stress", Scalar, "The von Mises stress", attr="_vm"),
        input("cell_dislocation_density", Scalar, "The cell dislocation density"),
        input("wall_dislocation_density", Scalar, "The wall dislocation density"),
        input("temperature", Scalar, "The temperature"),
        input("env_factor", Scalar, "The environment factor"),
        output("output_rate", Scalar, "The output rate", attr="_out_rate"),
        option("model_file_name", str, "The name of the model file", attr="_model_file"),
        option(
            "model_file_variable_name",
            str,
            "The name of the variable in the model file",
            attr="_out_name",
        ),
    )

    # Auto-declared by the schema (_store_schema_values). ``_vm`` is the resolved
    # von Mises stress variable name; the options hold the JSON file and the
    # selected output variable name.
    _vm: str
    _out_rate: str
    _model_file: str
    _out_name: str

    def __post_init__(self) -> None:
        with open(self._model_file) as f:
            j = json.load(f)

        # 1D grids (buffers) and per-axis input transforms, in grid-axis order.
        for i, key in enumerate(_GRID_KEYS):
            self.register_buffer(
                f"_grid_{i}", torch.tensor(self._require(j, key), dtype=torch.float64)
            )
        self._in_transforms = [self._read_transform(j, key) for key in _GRID_KEYS]

        # 6D grid values for the selected output (validate rectangularity first,
        # matching the v2 json_6Dvector_to_torch size check).
        values = self._require(j, self._out_name)
        self._check_rectangular(values, self._out_name)
        self.register_buffer("_grid_values", torch.tensor(values, dtype=torch.float64))

        # Output transform, keyed off the selected output variable.
        if self._out_name == "out_ep":
            self._out_transform = self._read_transform(j, "out_strain_rate")
        elif self._out_name == "out_cell":
            self._out_transform = self._read_transform(j, "out_cell_rate")
        elif self._out_name == "out_wall":
            self._out_transform = self._read_transform(j, "out_wall_rate")
        else:
            raise RuntimeError(
                "This ouput variable is not implemented, model_file_variable_name: "
                + str(self._out_name)
            )

        # Only the inelastic strain rate needs a first derivative, w.r.t. the von
        # Mises stress, for the radial-return solve (matches the v2 model).
        if self._out_name == "out_ep":
            self.request_AD(inputs=[self._vm])  # type: ignore[operator]

    def forward(  # type: ignore[override]
        self,
        equivalent_plastic_strain: Scalar,
        von_mises_stress: Scalar,
        cell_dislocation_density: Scalar,
        wall_dislocation_density: Scalar,
        temperature: Scalar,
        env_factor: Scalar,
        *nl_params: Scalar,
        v=None,
    ):
        # Transform each input onto its grid axis (raw-tensor boundary, matching
        # the ML-surrogate request_AD pattern: the grad-tracking leaf request_AD
        # swaps in for von_mises_stress is exactly ``.data``), then interpolate
        # and transform the result back.
        points = [
            self._transform(von_mises_stress.data, self._in_transforms[0]),
            self._transform(temperature.data, self._in_transforms[1]),
            self._transform(equivalent_plastic_strain.data, self._in_transforms[2]),
            self._transform(cell_dislocation_density.data, self._in_transforms[3]),
            self._transform(wall_dislocation_density.data, self._in_transforms[4]),
            self._transform(env_factor.data, self._in_transforms[5]),
        ]
        out = Scalar(self._transform(self._interpolate(points), self._out_transform))
        if v is None:
            return out

        # out_ep's first derivative (w.r.t. von_mises_stress) is supplied by
        # request_AD, so __call__ intercepts before forward ever sees ``v``. The
        # cell/wall instances reach here with a ``v`` seed; like the v2 model they
        # carry no input derivative, so the chain rule is a structural zero.
        return out, self.apply_chain_rule(v, self._out_rate, {}, output=out)

    # ------------------------------------------------------------------
    # Interpolation
    # ------------------------------------------------------------------

    def _interpolate(self, points: list[torch.Tensor]) -> torch.Tensor:
        """64-corner (2^6) multilinear interpolation over the rectilinear grid."""
        # Fetch the grid buffers fresh so they reflect the current device/dtype
        # after any .to() move.
        grids = [getattr(self, f"_grid_{a}") for a in range(6)]
        grid_values = cast(torch.Tensor, self._grid_values)

        left_idx: list[torch.Tensor] = []
        # ``fracs[axis] = (left_weight, right_weight)`` for the bracketing pair.
        fracs: list[tuple[torch.Tensor, torch.Tensor]] = []
        for grid, pt in zip(grids, points):
            n = grid.shape[0]
            # searchsorted returns the right index; -1 makes it the left. Clamp
            # to [0, n-2] so the bracketing pair stays in range (extrapolation
            # falls out of the fraction formula below).
            idx = torch.clamp(torch.searchsorted(grid, pt) - 1, 0, n - 2)
            left = grid[idx]
            right = grid[idx + 1]
            left_fraction = (right - pt) / (right - left)
            left_idx.append(idx)
            fracs.append((left_fraction, 1.0 - left_fraction))

        result = torch.zeros_like(points[0])
        for corner in itertools.product((0, 1), repeat=6):
            vertex = grid_values[tuple(left_idx[a] + corner[a] for a in range(6))]
            weight = fracs[0][corner[0]]
            for a in range(1, 6):
                weight = weight * fracs[a][corner[a]]
            result = result + vertex * weight
        return result

    # ------------------------------------------------------------------
    # Transforms (data <-> grid coordinate). ``params`` is (type, values).
    # ------------------------------------------------------------------

    @staticmethod
    def _transform(data: torch.Tensor, params: tuple[str, list[float]]) -> torch.Tensor:
        ttype, p = params
        if ttype == "COMPRESS":
            factor, compressor, original_min = p[0], p[1], p[2]
            d1 = torch.sign(data) * torch.abs(data * factor) ** compressor
            return torch.log10(1.0 + d1 - original_min)
        if ttype == "DECOMPRESS":
            factor, compressor, original_min = p[0], p[1], p[2]
            d1 = 10.0**data - 1.0 + original_min
            return torch.sign(d1) * torch.abs(d1) ** (1.0 / compressor) / factor
        if ttype == "LOG10BOUNDED":
            factor, lowerbound, upperbound, logmin, logmax = p[0], p[1], p[2], p[3], p[4]
            rng = upperbound - lowerbound
            return rng * (torch.log10(data + factor) - logmin) / (logmax - logmin) + lowerbound
        if ttype == "EXP10BOUNDED":
            factor, lowerbound, upperbound, logmin, logmax = p[0], p[1], p[2], p[3], p[4]
            rng = upperbound - lowerbound
            return 10.0 ** (((data - lowerbound) * (logmax - logmin) / rng) + logmin) - factor
        if ttype == "MINMAX":
            data_min, data_max, scaled_min, scaled_max = p[0], p[1], p[2], p[3]
            return (data - data_min) / (data_max - data_min) * (scaled_max - scaled_min) + scaled_min
        raise RuntimeError("Unrecognized transform: " + str(ttype))

    # ------------------------------------------------------------------
    # JSON readers
    # ------------------------------------------------------------------

    @staticmethod
    def _require(j: dict, key: str):
        if key not in j:
            raise RuntimeError(f"The key '{key}' is missing from the JSON data file.")
        return j[key]

    def _read_transform(self, j: dict, axis: str) -> tuple[str, list[float]]:
        return (
            self._require(j, axis + "_transform_type"),
            self._require(j, axis + "_transform_values"),
        )

    @staticmethod
    def _check_rectangular(data, key: str) -> None:
        """Validate the 6D grid is rectangular (every sub-list at a given depth
        matches the size of the first sub-list at that depth), matching the v2
        ``json_6Dvector_to_torch`` size check."""

        def walk(v):
            if not isinstance(v, list) or not v or not isinstance(v[0], list):
                return
            n = len(v[0])
            for x in v:
                if len(x) != n:
                    raise RuntimeError(f"Incorrect JSON interpolation grid size for '{key}'.")
                walk(x)

        walk(data)


__all__ = ["LAROMANCE6DInterpolation"]
